src/omp/backend.f90

```diff
@@ -379,18 +412,91 @@ subroutine vecadd_omp(self, a, x, b, y)
     class(field_t), intent(in) :: x
     real(dp), intent(in) :: b
     class(field_t), intent(inout) :: y
+    integer, dimension(3) :: dims
+    integer :: i, j, k, ii
+
+    integer :: nvec, remstart
+
+    if ((x%dir /= y%dir) .or. (x%data_loc /= y%data_loc)) then
+      error stop "Called vector add with incompatible fields"
+    end if
+
+    dims = size(x%data)
+    nvec = dims(1)/SZ
+    remstart = nvec*SZ + 1
 
-    y%data = a*x%data + b*y%data ! fixme
+    !$omp parallel do private(i, ii) collapse(2)
+    do k = 1, dims(3)
+      do j = 1, dims(2)
+        ! Execute inner vectorised loops
+        do ii = 1, nvec
+          !$omp simd
+          do i = 1, SZ
+            y%data(i + (ii - 1)*SZ, j, k) = &
+              a*x%data(i + (ii - 1)*SZ, j, k) + &
+              b*y%data(i + (ii - 1)*SZ, j, k)
+          end do
+          !$omp end simd
+        end do
+
+        ! Remainder loop
+        do i = remstart, dims(1)
+          y%data(i, j, k) = a*x%data(i, j, k) + b*y%data(i, j, k)
+        end do
+      end do
+    end do
+    !$omp end parallel do
 
   end subroutine vecadd_omp
 
   real(dp) function scalar_product_omp(self, x, y) result(s)
+
+    use mpi
+
     implicit none
 
     class(omp_backend_t) :: self
     class(field_t), intent(in) :: x, y
+    integer, dimension(3) :: dims
+    integer :: i, j, k, ii
+    integer :: ierr
+
+    integer :: nvec, remstart
+
+    if ((x%dir /= y%dir) .or. (x%data_loc /= y%data_loc)) then
+      error stop "Called scalar product with incompatible fields"
+    end if
+
+    dims = self%mesh%get_field_dims(x)
```

Comment:
> get_field_dims in the mesh class is a bit misleading I believe. It returns the actual dimensions for a DIR_C array, but for DIR_X/Y/Z arrays its a bit different. It returns the actual number of entries along its own DIR, but the first and last dims are always SZ and the number of groups.

I'm not sure I understand the use case for `get_field_dims` then, based on
```
 pure function get_padded_dims_dir(self, dir) result(dims_padded)
  !! Getter for padded dimensions with structure in `dir` direction
```
I would have thought the former gets the non-padded dimensions, and the latter the padded dimensions?

Reply:
I can't think of a use case for `get_field_dims` to be honest. If one needs non-padded dimensions along the present `DIR` and also the number of groups its better to get these with two separate calls instead.

Reply:
The problem is we can't really have non-padded dimensions in `DIR_X/Y/Z` arrangements. Certain combinations of the first index and last index result in padding in that case.

Reply:
Would it not be better to have it return the non-padded dimensions? Are the padded dimensions not just `size(phi%data)`? How do we get the non-padded dimensions then? I'm sorry this is unclear to me

Reply:
This is very tricky and not documented at all, my bad.
```
   2 4 6 8 10
   o o o o o
   x x x x x
   x x x x x
___x x x x x___
   x x x x x
   x x x x x
   x x x x x
   x x x x x
   1 3 5 7 9
```
group numbers are either below or above. Each group have 4 lines in total. `x` indicates a real entry, `o` indicates a padding. `_`is just a visual guide for separating two groups. Domain size is (nx=n, ny=7, nz=5) Lets say x-direction is towards inside the screen, y-dir is up, and z-dir is to the right. The dims of our `DIR_X` array shown here is `(SZ=4, nx, 10)`. The groups numbered 2, 4, 6, 8, 10 have some padding in them, but 1, 3, 5, 7, and 9 don't. So we can't really have a non-padded `k` index. What we can say is that only group numbers divisible by 2 include some amount of padding. To make this easy we can divide a `k` loop into two loops, so that one loop goes from 1 to 2(because we have two groups in y-direction, ny=7, SZ=4). and then the other loop goes from 1 to 5 (because nz=5).

Reply:
padded dimension of an array is equal to `size(phi%data)`.

By the way, the easiest solution here would be calling a reorder function and reorder whatever `DIR` we have happen to have in `scalar_product` into `DIR_C`. Then the 3 dimensions and their paddings are very easy to deal with. non-padded dimensions can be obtained by `get_dims(phi%data_loc)`, and then loop bounds will make sure we stay within the non-padded entries. The only issue with this is it does a reorder unnecessarily. In the solver we use `scalar_product` only once in a while, for the enstrophy, so no problem if it doens't perform well. But I think the iterative solver uses it a bit more often, but not sure how much weight it has overall.

Reply:
OK, I'll go with the reordering approach. The iterative solver shouldn't have to do this, we only need to provide a routine for `lapl(p)`, any scalar products in the solver are handled by PETSc

Reply:
Could we not have a function that returns the 'non padded' `SZ` value given the group index? It won't be perfect for performance because vectorisation might be harder, but that should be better than reordering and be more 'clear'

Reply:
OK, scalar product is now implemented using reordering to DIR_C :)